ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,294 @@
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  STM m ()
+demoteObjector handlesVar = do
+  handles <- Map.toList <$> readTVar handlesVar
+  case List.find (isObjector . cschJumping . snd) handles of
+    Just (objector, ChainSyncClientHandle {cschJumping = Objector intersection}) ->
+      setJumpingState handlesVar objector =<< newJumper intersection FoundIntersection
+    _ -> pure ()
+  where
+    isObjector (Objector _) = True
+    isObjector _            = False
+
+newJumper ::
+  ( MonadSTM m,
+    HasHeader blk
+  ) =>
+  Point blk ->
+  ChainSyncJumpingJumperState blk ->
+  STM m (ChainSyncJumpingState m blk)
+newJumper intersection state = do
+  nextJumpVar <- newTVar Nothing
+  pure $ Jumper nextJumpVar intersection state
+
+-- | Register a new ChainSync client to the given map of handles. If there is no
+-- dynamo, then it starts as a dynamo; otherwise, it starts as a jumper.
+registerClient ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  -- | A function to make a client handle from a jumping state.
+  (ChainSyncJumpingState m blk -> ChainSyncClientHandle m blk) ->
+  STM m ()
+registerClient handlesVar peer mkHandle = do
+  cschJumping <-
+    getDynamo handlesVar >>= \case
+      Nothing -> pure $ Dynamo Origin
+      Just _ -> newJumper GenesisPoint Happy
+  modifyTVar handlesVar $ Map.insert peer $ mkHandle cschJumping
+
+unregisterClient ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  STM m ()
+unregisterClient handlesVar peer = do
+  handle <- (Map.! peer) <$> readTVar handlesVar
+  modifyTVar handlesVar $ Map.delete peer
+  case cschJumping handle of
+    Jumper _ _ _ -> pure ()
+    Objector _   -> electNewObjector handlesVar
+    Dynamo _     -> electNewDynamo handlesVar
+
+-- | Look into all objector candidates and promote the one with the oldest
+-- intersection with the dynamo as the new objector.
+electNewObjector ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  STM m ()
+electNewObjector handlesVar = do
+  demoteObjector handlesVar
+  -- Get all the jumpers that disagree with the dynamo and for which we know the
+  -- precise intersection; find the one with the oldest intersection (which is
+  -- very important) and promote it to objector.
+  sortedObjectorCandidates <-
+    List.sortOn (pointSlot . snd)
+      . mapMaybe
+        ( \(peer, handle) ->
+            case cschJumping handle of
+              Jumper _ intersection FoundIntersection -> Just (peer, intersection)
+              _ -> Nothing
+        )
+      . Map.toList
+      <$> readTVar handlesVar
+  case sortedObjectorCandidates of
+    [] -> pure ()
+    (objector, intersection) : _ ->
+      setJumpingState handlesVar objector $ Objector intersection
```

Comment:
What is the effect of ignoring the jumpers that haven't found the intersection yet?

Is LoP leaking paused for them? If the bucket is leaking, they must accept jumps quickly for the candidate fragment to be updated.

Are they considered as idling? If so, they also must accept jumps quickly so GDD does not disconnect them when they disagree by more than k slots when the dynamo or the objector move ahead.

Reply:
> What is the effect of ignoring the jumpers that haven't found the intersection yet? I'm guessing they will hold back LoE until something disconnects them.

Indeed, we know that they are unhappy with the dynamo's fragment so there is an intersection somewhere and the LoE should prevent us from moving forward. I think we have to ignore them because we don't know what to do with them yet: there is no point asking them to jump to the dynamo's fragment because we already know they disagree but we also don't have enough information yet to decide if they are the objector whose intersection with the dynamo is the oldest.

The key point here is that we need to make sure that the objector whose intersection with the dynamo is the oldest is running ChainSync; otherwise, the whole system might block because the LoE cannot move forward (because of them) but then the GDD might not have enough information to kill the other peers. In the current implementation with at most one objector, that means that we need to make sure that's the one that's running. We could imagine other designs (eg. where all unhappy jumpers start ChainSync, or maybe we classify them by answers to the jump requests, etc.).

Note that for now we try to find the exact intersection while we in fact only would need to pinpoint it down enough to decide that they are the oldest!

> Is LoP leaking paused for them? If the bucket is leaking, they must accept jumps quickly for the candidate fragment to be updated. But they never earn tokens, because they don't request headers.
> 
> Are they considered as idling? If so, they also must accept jumps quickly so GDD does not disconnect them when they disagree by more than k slots when the dynamo or the objector move ahead.

The ChainSync clients are indeed idling when waiting for the next instruction from ChainSync jumping.

> If no Genesis mechanism is active to disconnect them, they can delay each jump request by as much as the intersect timeout.

Absolutely, so 10s per jump request, which is a lot. Finding the intersection as it is done currently can require quite a few back-and-forth with `MsgFindIntersect` and they can all take a while. Again, the idea is that then we'll kill the peer that disagrees so it's not too bad to have paid even a few minutes to decide; the rest of the sync should be fast. However, it would be easy for an adversary to try and leash the sync by serving the honest chain but waiting 9.9s to answer each jump request. I think this pushes for a value of `jumpSize` small enough that the dynamo still has at least 10s worth of work to do.

Reply:
If all jumpers are idling, they will be disconnected by GDD when the dynamo or the objector move more than k blocks ahead.

Are there any guarantees that jumpers corresponding to honest peers will have a fair chance to find intersections before being disconnected?

Also, when intersections are finally found, I'm not seeing any code to update the candidate fragments of those jumpers.

Reply:
@amesgen suggested considering jumpers as not idling and using short timeouts for MsgFindIntersect.

Perhaps ensuring safety and liveness is possible, but we need extra provisions,
* Set the candidate fragments when intersections are found, so the LoE can move forward promptly on prefixes where there is agreement.
* Hold objectors accountable for their earlier objections. If they object when jumping, but yield the honest chain when acting as objectors, GDD and LoP won't be able to disconnect anyone.
* Select a new objector when the current objector requires headers after the forecast horizon. This can happen when the intersection of the objector is much later than the LoE tip.

Reply:
> Hold objectors accountable for their earlier objections.

After more discussion I offer the following fix.

If the first header of the objector after the intersection agrees with the dynamo and the slot is inside the last genesis window, move the objector to a new state `NoCSJer` to which no jumps are offered, and from which headers are downloaded via ChainSync as if CSJ was disabled.

If the slot is before the last genesis window, disconnect the peer, as the rollback is too deep.

CSJ should be turned off when no jumpers are left.

Reply:
Following the CSJ description, this measure is not necessary. Proscribing historical rollbacks might be helpful, but it is an optimization, if we elect new objectors for as long as GDD is inconclusive.